src/data/finsupp/lattice.lean

```diff
@@ -0,0 +1,119 @@
+    intro s, apply and_iff_right_of_imp, rw ← not_and_distrib, rw not_imp_not,
+    intro h1, rw h1.left, rw h1.right, apply h, }, }
+
+lemma binary_op_pointwise_apply {f : β → β → β} (h : f 0 0 = 0) (a b : α →₀ β) (s : α) :
+  (binary_op_pointwise h a b) s = f (a s) (b s) := rfl
+
+instance [semilattice_inf β] : semilattice_inf (α →₀ β) :=
+{ inf := binary_op_pointwise inf_idem,
+  inf_le_left := by { intros, rw le_def, intro, apply inf_le_left, },
+  inf_le_right := by { intros, rw le_def, intro, apply inf_le_right, },
+  le_inf := by { intros, rw le_def at *, intro, apply le_inf (a_1 s) (a_2 s) },
+  ..finsupp.partial_order,
+}
+
+@[simp]
+lemma inf_apply [semilattice_inf β] {a : α} {f g : α →₀ β} : (f ⊓ g) a = f a ⊓ g a := rfl
+
+lemma support_inf {f g : α →₀ γ} : (f ⊓ g).support = f.support ∩ g.support :=
+begin
+  change (binary_op_pointwise inf_idem f g).support = f.support ∩ g.support,
+  unfold binary_op_pointwise, dsimp, ext,
+  simp only [mem_support_iff,  ne.def,
+    finset.mem_union, finset.mem_filter, finset.mem_inter],
+  split; intro h, cases h with i j, split; intro con; rw con at j; apply j; simp [zero_le],
+  { split, left, exact h.left, rw inf_eq_min, unfold min, split_ifs; tauto, }
+end
+
+instance [semilattice_sup β] : semilattice_sup (α →₀ β) :=
+{ sup := binary_op_pointwise sup_idem,
+  le_sup_left := by { intros, rw le_def, intro, apply le_sup_left, },
+  le_sup_right := by { intros, rw le_def, intro, apply le_sup_right, },
+  sup_le := by { intros, rw le_def at *, intro, apply sup_le (a_1 s) (a_2 s) },
+  ..finsupp.partial_order,
+}
+
+@[simp]
+lemma sup_apply [semilattice_sup β] {a : α} {f g : α →₀ β} : (f ⊔ g) a = f a ⊔ g a := rfl
+
+@[simp]
+lemma support_sup
+  {f g : α →₀ γ} : (f ⊔ g).support = f.support ∪ g.support :=
+begin
+  change (binary_op_pointwise sup_idem f g).support = f.support ∪ g.support,
+  unfold binary_op_pointwise, dsimp, rw finset.filter_true_of_mem,
+  intros x hx, rw ← bot_eq_zero, rw sup_eq_bot_iff,
+   revert hx,
+  simp only [not_and, mem_support_iff, bot_eq_zero, ne.def, finset.mem_union], tauto,
+end
+
+instance lattice [lattice β] : lattice (α →₀ β) :=
+{ .. finsupp.semilattice_inf, .. finsupp.semilattice_sup}
+
+instance semilattice_inf_bot : semilattice_inf_bot (α →₀ γ) :=
+{ ..finsupp.order_bot, ..finsupp.lattice}
+
+lemma of_multiset_strict_mono : strict_mono (@finsupp.of_multiset α) :=
+begin
+  unfold strict_mono, intros, rw lt_iff_le_and_ne at *, split,
+  { rw finsupp.le_iff, intros s hs, repeat {rw finsupp.of_multiset_apply},
+    rw multiset.le_iff_count at a_1, apply a_1.left },
+  { have h := a_1.right, contrapose h, simp at *,
+    apply finsupp.equiv_multiset.symm.injective h }
+end
+
+lemma bot_eq_zero : (⊥ : α →₀ γ) = 0 := rfl
+
+@[simp]
+lemma disjoint_iff {x y : α →₀ γ} : disjoint x y ↔ disjoint x.support y.support :=
+begin
+  unfold disjoint, repeat {rw le_bot_iff},
+  rw [finsupp.bot_eq_zero, ← finsupp.support_eq_empty, finsupp.support_inf], refl,
+end
```

Comment:
If this really should exist, it probably should live in `data.finsupp`. I'm dubious about it as a simp lemma.

Reply:
I can desimp it, but it's defined in terms of the lattice inf.